README.md
[![CircleCI](https://circleci.com/gh/jahuty/jahuty-php.svg?style=svg)](https://circleci.com/gh/jahuty/jahuty-php) [![codecov](https://codecov.io/gh/jahuty/jahuty-php/branch/master/graph/badge.svg?token=XELPI4FWMI)](https://codecov.io/gh/jahuty/jahuty-php)

# jahuty-php

Welcome to the PHP SDK for [Jahuty's API](https://docs.jahuty.com/api)!

## Installation

This library requires [PHP 7.3+](https://secure.php.net).

It is multi-platform, and we strive to make it run equally well on Windows, Linux, and OSX.

It should be installed via [Composer](https://getcomposer.org). To do so, add the following line to the `require` section of your `composer.json` file, and run `composer update`:

```javascript
{
   "require": {
       "jahuty/jahuty-php": "^5.4"
   }
}
```

## Usage

Instantiate the library with your [API key](https://docs.jahuty.com/api#authentication) and use the `snippets->render()` method to render your snippet:

```php
$jahuty = new \Jahuty\Client('YOUR_API_KEY');

$render = $jahuty->snippets->render(YOUR_SNIPPET_ID);
```

Then, output the render's content by casting it to a `string` or using its `getContent()` method:

```php
$jahuty = new \Jahuty\Client('YOUR_API_KEY');

$render = $jahuty->snippets->render(YOUR_SNIPPET_ID);

echo $render;
echo $render->getContent();
```

In an HTML view:

```html+php
<?php
  $jahuty = new \Jahuty\Client('YOUR_API_KEY');
?>
<!doctype html>
<html>
<head>
    <title>Awesome example</title>
</head>
<body>
    <?php echo $jahuty->snippets->render(YOUR_SNIPPET_ID); ?>
</body>
```

You can also use [tags](https://docs.jahuty.com/components/tags) to select a collection of snippets with the `snippets->allRenders()` method:

```php
$jahuty = new \Jahuty\Client('YOUR_API_KEY');

$renders = $jahuty->snippets->allRenders('YOUR_TAG');

foreach ($renders as $render) {
  echo $render;
}
```

## Content versions

By default, this library will render a snippet's _published_ content. If you'd like to render its latest _staged_ content instead, you can use the `prefer_latest` configuration option at the client or render level:

```php
// Prefer the latest content for all renders.
$jahuty = new \Jahuty\Client('YOUR_API_KEY', ['prefer_latest' => true]);
```

```php
// Use the default published content for all renders.
$jahuty = new \Jahuty\Client('YOUR_API_KEY');

// And, render the latest content for this one.
$render = $jahuty->snippets->render(YOUR_SNIPPET_ID, ['prefer_latest' => true]);
```

## Parameters

You can pass [parameters](https://docs.jahuty.com/liquid/parameters) into your renders using the `params` configuration option:

```php
$jahuty = new \Jahuty\Client('YOUR_API_KEY');

$render = $jahuty->snippets->render(YOUR_SNIPPET_ID, [
  'params' => [
    'foo' => 'bar'
  ]
]);
```

The parameters above would be equivalent to [assigning the variables](https://docs.jahuty.com/liquid/variables) below in your snippet:

```html
{% assign foo = "bar" %}
```

If you're rendering a collection of snippets, the first dimension of the `params` key determines the parameters' scope. Use the asterisk key (`*`) to pass the same parameters to all snippets, or use the snippet id as key to pass parameters to a specific snippet.

```php
$jahuty = new \Jahuty\Client('YOUR_API_KEY');

$renders = $jahuty->snippets->allRenders('YOUR_TAG', [
  'params' => [
    '*' => [
      'foo' => 'bar'
    ],
    1 => [
      'baz' => 'qux'
    ]
  ]
]);
```

The two parameter lists will be recursively merged, and the snippet's parameters will take precedence. In the example below, the variable `foo` will be assigned the value `"bar"` for all snippets, except for snippet 1, where it will be assigned the value `"qux"`:

```php
$jahuty = new \Jahuty\Client('YOUR_API_KEY');

$renders = $jahuty->snippets->allRenders('YOUR_TAG', [
  'params' => [
    '*' => [
      'foo' => 'bar'
    ],
    1 => [
      'foo' => 'qux'
    ]
  ]
]);
```

## Tracking renders

You can use the `render()` method's `location` configuration option to report the absolute URL where a snippet is being rendered. This helps your team preview their changes, and it helps you find and replace deprecated snippets.

```php
$jahuty = new \Jahuty\Client('YOUR_API_KEY');

$render = $jahuty->snippets->render(YOUR_SNIPPET_ID, [
  'location' => 'https://example.com'
]);
```

Note, locations are only reported when a request is sent to Jahuty's API. As a result, locations will not be reported in some scenarios. For example, if the call to `render()` results in a cache hit, a request will not sent to Jahuty's API, and the location will not be reported.

This configuration option is not supported by the `allRenders()` method.

## Caching

Caching controls how frequently this library requests content from Jahuty's API.

* In _development_, where content is frequently changing and low traffic, you can use the default in-memory store to view content changes instantaneously.
* In _production_, where content is more stable and high traffic, you can configure persistent caching to reduce network requests and improve your application's response time.

### Caching in memory (default)

By default, Jahuty uses an in-memory cache to avoid requesting the same render more than once during the same request lifecycle. For example:

```php
$jahuty = new \Jahuty\Client('YOUR_API_KEY');

// This call sends a synchronous API request; caches the result in memory; and,
// returns the resulting render.
$render1 = $jahuty->snippets->render(YOUR_SNIPPET_ID);

// This call skips sending an API request and uses the cached render instead.
$render2 = $jahuty->snippets->render(YOUR_SNIPPET_ID);
```

The in-memory cache only persists for the duration of the original request, however. At the end of the request's lifecycle, the cache will be discarded. To store renders across requests, you need a persistent cache.

### Caching persistently

A persistent cache allows renders to be cached across multiple requests. This reduces the number of synchronous network requests to Jahuty's API and improves your application's response time.

To configure a persistent cache, pass a [PSR-16 `CacheInterface`](https://www.php-fig.org/psr/psr-16/) implementation to the client via the `cache` configuration option:

```php
$jahuty = new \Jahuty\Client('YOUR_API_KEY', ['cache' => CACHE_INSTANCE]);
```

The persistent cache implementation you choose and configure is up to you. There are many libraries available, and most frameworks provide their own. Any PSR-16 compatible implementation will do.

### Expiring

There are three methods for configuring a render's Time to Live (TTL), the amount of time between when a it is stored and when it's considered stale. From lowest-to-highest precedence, the methods are:

1. configuring your caching implementation,
1. configuring this library's default TTL, and
1. configuring a render's TTL.

#### Configuring your caching implementation

You can usually configure your caching implementation with a default TTL. If no other TTL is configured, this library will defer to the caching implementation's default TTL. How exactly you configure this depends on the library you choose.

#### Configuring this library's default TTL

You can configure a default TTL for all of this library's renders by passing an integer number of seconds or a `DateInterval` instance via the client's `ttl` configuration option:

```php
// Cache all renders for sixty seconds.
$jahuty = new \Jahuty\Client('YOUR_API_KEY', [
  'cache' => $cache,
  'ttl'   => 60
]);
```

If this library's default TTL is set, it will take precedence over the default TTL of the caching implementation.

#### Configuring a render's TTL

You can configure one render's TTL by passing an integer number of seconds or a `DateInterval` instance via its `ttl` configuration option:

```php
// Use the caching implementation's TTL for all renders.
$jahuty = new \Jahuty\Client('YOUR_API_KEY', ['cache' => $cache]);

// But, cache this render for 60 seconds.
$render = $jahuty->snippets->render(1, [
  'ttl' => 60
]);
```

If a render's TTL is set, it will take precedence over the library's default TTL and the caching implementation's TTL.

### Caching collections

By default, this library will cache each render returned by `allRenders()`:

```php
$jahuty = new \Jahuty\Client('YOUR_API_KEY');

// This call sends a network request, caches each render, and returns the
// collection.
$jahuty->snippets->allRenders('YOUR_TAG');

// ... later in your application

// If this snippet exists in the previously rendered collection, the cached
// render will be used.
$render = $jahuty->snippets->render(YOUR_SNIPPET_ID);
```

This is a powerful feature. By using tags and the `allRenders()` method, you can render and cache the content of an entire application with a single network request.

Further, when `allRenders()` can be called periodically outside your request cycle (e.g., in a background job), you can turn your persistent cache into your content storage mechanism. You can render and cache your dynamic content as frequently as your like without any hit to your application's response time.

### Disabling caching

You can disable caching, even the default in-memory caching, by passing a `ttl` of zero (`0`) or a negative integer (e.g., `-1`) via any of the methods described above. For example:

```php
// Disable all caching
$jahuty1 = new \Jahuty\Client('YOUR_API_KEY', ['ttl' => 0]);

// Disable caching for this render
$jahuty2 = new \Jahuty\Client('YOUR_API_KEY');
$jahuty2->snippets->render(1, ['ttl' => 0]);
```

## Errors

If Jahuty's API returns any [status code](https://docs.jahuty.com/api) other than `2xx`, a `Jahuty\Exception\Error` will be thrown:

```php
try {
  $jahuty = new \Jahuty\Client('YOUR_API_KEY');
  $render = $jahuty->snippets->render(YOUR_SNIPPET_ID);
} catch (\Jahuty\Exception\Error $e) {
  // The API returned something besides 2xx status code. Access the problem
  // object for more information.
  $problem = $e->getProblem();

  echo $problem->getStatus();  // returns status code
  echo $problem->getType();    // returns URL to more information
  echo $problem->getDetail();  // returns description of error
}
```

That's it!

## License

This library is licensed under the [MIT license](LICENSE).

## Contributing

Contributions are welcome! If you spot an error, please open an issue or [let us know](https://www.jahuty.com/contact).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [5.4.0] - 2020-07-28

### Added

- A `location` option has been added to the `render` method, which allows you to report the absolute URL where the snippet is being rendered.
- A `prefer_latest` configuration option has been added to replace the intentional but verbose `prefer_latest_content` configuration option.

### Deprecated

- The `prefer_latest_content` configuration option is deprecated in favor of the `prefer_latest` configuration option.

### Fixed

- Fixed [#27](https://github.com/jahuty/jahuty-php/issues/27), where the content version was not considered when fetching renders from the cache.

## 5.3.0 - 2020-04-27

- Add the `prefer_latest_content` configuration setting to the client and render methods. This setting allows you to render a snippet's _latest_ content instead of the default _published_ content.

## 5.2.1 - 2020-03-16

- Add system tests for collections, parameterized snippets, and problems.

## 5.2.0 - 2020-03-05

- Add support for rendering and caching a collection of snippets via the `snippets->allRenders()` method.
- Remove the `Cache\Manager` and `Service\Factory` and move caching to the `Service\Snippet`. We got too fancy, and it got in the way.

## 5.1.1 - 2020-12-27

- Move `Jahuty\Ttl\Ttl` to `Jahuty\Cache\Ttl`.
- Change the caching section introduction and examples in the `README` to improve explanations.

## 5.1.0 - 2020-12-26

- Add a second, optional parameter to `Jahuty\Client`, an options array with support for the following options: `cache`, a `CacheInterface` implementation); `ttl`, a `null`, `int`, or `DateInterval` SDK-wide time-to-live; and, `base_uri`, the base URI for API requests (useful for testing).
- Add `fetch()` to `Jahuty\Client`, which checks the cache before requesting an action.
- Add `Jahuty\Cache\Memory`, a PSR-16 compliant in-memory cache adapter.
- Change `Jahuty\Client` to default to the in-memory cache, if no `cache` is configured.
- Move system tests to a separate test case, and improve the testing of `Jahuty\Client`.

## 5.0.0 - 2020-12-23

- Reduce root namespace from `Jahuty\Jahuty` to `Jahuty`.
- Change from a static-based architecture (e.g., `\Jahuty\Jahuty\Snippet::render(1)`) to an instance-based one (e.g., `$jahuty->snippets->render(1)`) to make the library easier to develop, test, and use.
- Fix links in `README`.
- Add code coverage analysis.

## 4.0.0 - 2020-07-04

- Rename `Serivce\Get` to `Service\Render`.
- Rename `Snippet::get` to `Snippet::render`.
- Change the optional second argument of `render()` from an array of `params` to an options array with a `params` key.
- Rename `Data\Snippet` to `Data\Render`.
- Remove `id` attribute from `Data\Render`.
- Change API endpoint from `snippets/:id` to `snippets/:id/render`.

## 3.1.1 - 2020-03-15

- Change snippet parameters to JSON query string parameter (e.g., `params={"foo":"bar"}`) from serialized query string parameter (e.g.,`params[foo]=bar`).

## 3.1.0 - 2020-03-13

- Add snippet parameters.
- Add `Jahuty::getClient()` to memoize Guzzle's HTTP client.
- Remove `Request` object and move constants like `HEADERS` and `BASE_URI` to `Jahuty`.

## 3.0.0 - 2020-03-08

- Add `Jahuty\Jahuty\Jahuty` to store API key and current version number.
- Remove `Jahuty\Jahuty\Snippet::key()` method.

## 2.0.0 - 2020-03-08

- Change namespace from `Jahuty\Snippet` to `Jahuty\Jahuty`.
- Change repository and package name from `jahuty/snippets-php` to `jahuty/jahuty-php`.

## 1.0.0 - 2019-09-02

- Initial release


src/Client.php
<?php

namespace Jahuty;

use Psr\SimpleCache\CacheInterface;

/**
 * Sends requests to Jahuty's API.
 *
 * @property  Service\Snippet  $snippets
 */
class Client
{
    private $apiClient;

    private $apiKey;

    private $baseUri = Jahuty::BASE_URI;

    private $cache;

    private $preferLatest = false;

    private $requests;

    private $responses;

    private $services = [];

    private $ttl;

    public function __construct(string $apiKey, array $options = [])
    {
        $this->apiKey = $apiKey;

        $this->cache = new Cache\Memory();
        $this->ttl   = new Cache\Ttl(null);

        $this->unpackOptions($options);
    }

    public function __get(string $name): Service\Service
    {
        if ($name !== 'snippets') {
            throw new \OutOfBoundsException("Service '$name' not found");
        }

        if (!\array_key_exists($name, $this->services)) {
            $this->services[$name] = new Service\Snippet(
                $this,
                $this->cache,
                $this->ttl,
                $this->preferLatest
            );
        }

        return $this->services[$name];
    }

    /**
     * Executes an action against the API and returns result.
     *
     * @param   Action\Action  $action
     * @return  array|Resource\Resource
     */
    public function request(Action\Action $action)
    {
        if (null === $this->requests) {
            $this->requests = new Request\Factory($this->baseUri);
        }

        $request = $this->requests->new($action);

        if (null === $this->apiClient) {
            $this->apiClient = new Api\Client($this->apiKey);
        }

        $response = $this->apiClient->send($request);

        if (null === $this->responses) {
            $this->responses = new Response\Handler();
        }

        $result = $this->responses->handle($action, $response);

        if ($result instanceof Resource\Problem) {
            throw new Exception\Error($result);
        }

        return $result;
    }

    public function setBaseUri(string $baseUri): self
    {
        $this->baseUri = $baseUri;

        return $this;
    }

    public function setCache(CacheInterface $cache): self
    {
        $this->cache = $cache;

        return $this;
    }

    /**
     * @deprecated  use preferLatest() instead (2020-07-28)
     */
    public function setPreferLatestContent(bool $preferLatestContent): self
    {
        return $this->setPreferLatest($preferLatestContent);
    }

    public function setPreferLatest(bool $preferLatest): self
    {
        $this->preferLatest = $preferLatest;

        return $this;
    }

    public function setTtl($ttl): self
    {
        $this->cache = new Cache\Ttl($ttl);

        return $this;
    }

    /**
     * Unpacks the options array into object properties.
     *
     * @param   array  $options
     *   @option  string  base_uri  the base uri of API requests (optional; if
     *     omitted, defaults to the value of Jahuty::BASE_URI constant)
     *   @option  CacheInterface  cache  the client's cache (optional; if
     *     ommitted, defaults to in-memory cache)
     *   @option  int|DateTime  ttl  the default time-to-live when writing
     *     to the cache (optional; if omitted, uses the method's $ttl argument
     *     or the cache's default setting, in that order)
     *   @option  bool  prefer_latest_content  a flag indicating whether or not
     *     to prefer the latest content version (optional; if omitted, defaults
     *     to published content version) (deprecated)
     *   @option  bool  prefer_latest  a flag indicating whether or not to
     *     prefer the latest content version (optional; if omitted, defaults
     *     to published content version)
     * @return  void
     */
    private function unpackOptions(array $options): void
    {
        if (isset($options['base_uri'])) {
            $this->setBaseUri($options['base_uri']);
        }

        if (isset($options['cache'])) {
            $this->setCache($options['cache']);
        }

        if (isset($options['ttl'])) {
            $this->setTtl($options['ttl']);
        }

        if (isset($options['prefer_latest_content']) || isset($options['prefer_latest'])) {
            $this->setPreferLatest($options['prefer_latest_content'] || $options['prefer_latest']);
        }
    }
}
